### PR TITLE
new functionalities and api change for xr-transformer

### PR DESCRIPTION
**Pull request by jiong-zhang**
*Issue #, if available:* NA

*Description of changes:*
 * Enable cost-sensitive linear model training in `TransformerMatcher`.
 * Unify bootstrapping params into `bootstrap_method`.
 * Add functionality in `TransformerMatcher` to override `pred_params` during prediction.
 * Add functionality to determin device (CPU or GPU) for projection layer by output size.
 * Add functionality to *inherit* the projection layer rather than using linear bootstrapping for *xr-transformer*.
 * Add functionality to choose different learning rate schedule for *xr-transformer*.
 * Add functionality to scale the number of fine-tuning steps for different sub-tasks in *xr-transformer*.
 * Fix inconsistent help doctoring for `nr-splits` of `xlinear` and `xtransformer`.
 * Add temporary directory to store encoded text tensors, this will save the trouble of repetitive text encoding when `saved-trn-pt` and `saved-val-pt` are not given.
 * Change of several default hyper-parameters in `xtransformer.train`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pecos/xmc/xtransformer/network.py [hunk 1]
```diff
@@ -124,19 +124,32 @@ def random_init(self):
...
+                Cp (float): the weight on positive samples. Default 100.0
+                Cn (float): the weight on negative samples. Default 100.0
+                threshold (float): the threshold to sparsify the model
         """
-        prob = MLProblem(X_emb, Y, C=C if M is not None else None, M=M)
         # use large Cp and Cn to reduce regularization
-        mat = MLModel.train(prob, threshold=0.0, Cp=100.0, Cn=100.0)
+        Cp = kwargs.get("Cp", 100.0)
+        Cn = kwargs.get("Cn", 100.0)
+        threshold = kwargs.get("threshold", 0)
```
Comment (OctoberChang) [inline pecos/xmc/xtransformer/network.py:152]: Maybe add one sentence say why threshold=0.0, and when will this cause OOM? Similarly, the W.toarray() operation.

File: pecos/xmc/xtransformer/matcher.py [hunk 2]
```diff
@@ -1316,13 +1362,23 @@ def train(
...
+                bootstrap_prob = MLProblem(
+                    init_embeddings,
+                    prob.Y,
+                    C=prob.C if prob.M is not None else None,
+                    M=prob.M,
+                    R=prob.Y if "weighted" in train_params.loss_function else None,
+                )
+                matcher.text_model.bootstrap(bootstrap_prob)
+                LOGGER.info("Initialized transformer text_model with xlinear!")
+            elif train_params.bootstrap_method == "inherit":
```
Comment (OctoberChang) [inline pecos/xmc/xtransformer/matcher.py:1379]: In practice, do we ever use this "inherit" to initialize the label embedding matrix?

Comment (jiong-zhang) [reply pecos/xmc/xtransformer/matcher.py:1379]: Yes, this will save the effort of training a linear model during bootstrapping and still have faster convergence than random initialization.

File: pecos/xmc/xtransformer/network.py (hunk 1 above)
Comment (jiong-zhang) [reply pecos/xmc/xtransformer/network.py:152]: The input feature matrix is always dense and therefore the sparsification is turned off. For the same reason, `W.toarray()` will not introduce additional memory cost as `W` is already a dense matrix. I'll add some comments here.

**Review by OctoberChang — APPROVED** [review]
